packages/simulation/src/convert.ts

```diff
@@ -0,0 +1,647 @@
+      if (initialTransitionTable[curElem][curStateID][1] === undefined) {
+        let sameStateTransitionExists = false
+        const checkStateID = initialTransitionTable[curElem][curStateID][0]
+        for (let i = 0; i < initialTransitionTable[checkStateID].length; i++) {
+          if (initialTransitionTable[checkStateID][i][0] === checkStateID) {
+            sameStateTransitionExists = true
+          }
+        }
+        if (!sameStateTransitionExists) {
+          removeTheseStates.push(checkStateID)
+        }
+        initialTransitionTable[curElem].splice(curStateID, 1)
+      }
+    }
+  }
+  for (let curElem = 0; curElem < removeTheseStates.length; curElem++) {
+    delete initialTransitionTable[removeTheseStates[curElem]]
+  }
+  return [initialTransitionTable, curInitialState, curFinalStates]
+}
+
+// This will mimic STEP 2 of the procedure by ensuring that for every DFA state, there is a symbol transition coming from it
+export function createSymbolFromEveryState (initialTransitionTable: {[key: StateID]: [StateID, ReadSymbol][]}, symbolsPresent: Set<ReadSymbol>, numberOfNFAStates: StateID, nfaGraph: FSAGraphIn) {
+  const symbolsArray = Array.from(symbolsPresent)
+  let nextAvailableStateID = numberOfNFAStates
+  // Go through all the original states defined and see if new states need to be created
+  for (let stateID = 0; stateID < numberOfNFAStates; stateID++) {
+    // If the state has been removed then ignore it and go to the next one
+    if (initialTransitionTable[stateID] === undefined) {
+      continue
+    } else {
+      // If the state has all the transitions then no need to do anything, go to next state
+      if (hasAllSymbols(initialTransitionTable[stateID], symbolsPresent)) {
+        continue
+      } else {
+        // If there isn't a transition going from a particular state, a new state isn't required, it can just transition to itself and still be
+        // a trap state. Not strictly required but it will make everything look nicer.
+        const hasTransitions = initialTransitionTable[stateID].length > 0
+        if ((!hasTransitions && !nfaGraph.states[stateID].isFinal)) {
+          for (let curElem = 0; curElem < symbolsArray.length; curElem++) {
+            initialTransitionTable[stateID].push([stateID, symbolsArray[curElem]])
+          }
+        } else if (!hasTransitions && nfaGraph.states[stateID].isFinal) {
+        // Else if the state is a final state but still needs transitions, it can just put all those transitions into one trap state
+          initialTransitionTable[nextAvailableStateID] = []
+          // New trap state
+          for (let curElem = 0; curElem < symbolsArray.length; curElem++) {
+            initialTransitionTable[nextAvailableStateID].push([nextAvailableStateID, symbolsArray[curElem]])
+          }
+          // Transitions to the trap state
+          for (let curElem = 0; curElem < symbolsArray.length; curElem++) {
```

Comment:
There are a lot of loops like this, but the index is only used to access the array, why not just use a `for of` loop?